### DISCUSSION TITLE
Is OpenVSwitch an obligatory dependency?

### DISCUSSION THREAD

**Discussion by pevogam** [avocado-framework/avocado-vt#3720] [Q&A] [answered]
Hi all, we have used Avocado VT for years now and never with OpenVSwitch which seems obligatory in here:
```python
mandatory_programs = {'qemu': basic_program_requirements + ['gcc'],
                      'spice': basic_program_requirements + ['gcc'],
                      'libvirt': basic_program_requirements,
                      'openvswitch': basic_program_requirements,
```
Nevertheless, bootstrapping Avocado VT never failed for us and all we got was a rather annoying error message every few lines like
```python
    def __init_openvswitch():
             if (not __ovs.check()):
                 raise Exception("Check of OpenVSwitch failed.")
         except Exception as e:
             LOG.error("Host does not support OpenVSwitch: %s", e)
             __ovs = None
     return __ovs
```
which never affected any of our VT tests. I was thinking of proposing a pull request and trying to drop or reduce the scope of the message but at present I am not sure if we are not missing out something with this dependency we don't use. Do you think it is indeed a mandatory dependency that should work and be used in all cases?

**Comment by pevogam** [comment]
@luckyh I have separated this possibly optional import in a discussion to try and figure out if it could be included in a pull request to reduce its verbosity too. Do you happen to have an opinion whether the verbosity here could also be reduced?

**Reply by luckyh** [reply]
@pevogam thank you for bringing up this topic, I'll leave my comments against some of the points.

> we have used Avocado VT for years now and never with OpenVSwitch which seems obligatory in here:

Well, in short, openvswitch is not mandatory for the majority of VT jobs except the testing of openvswitch tools (usually, when `vt.type == openvswitch`). For the code snippet you referred, it doesn't even check on whether the ovs bins was installed during the bootstrap stage.

https://github.com/avocado-framework/avocado-vt/blob/8181dfc0a47e88a7bbe418750e09dccb4337d901/virttest/bootstrap.py#L142

> all we got was a rather annoying error message every few lines like ... which never affected any of our VT tests.

However, some bridge relevant functionalities/procedures would perform the check and give such a warning upon failures due to the not well-defined behavior. At past, I sent [this commit](https://github.com/avocado-framework/avocado-vt/pull/3396/commits/5832a27a0e0afaedb358ff1975dac504c39877e8) to reduce the verbosity, however, it seems not all of them were suppressed.

So would you mind to provide the stack trace to help me know that where you see these ambiguous messages?

(p.s. to get the stack trace easily)
```patch
diff --git a/virttest/utils_net.py b/virttest/utils_net.py
index 5db2e6ea..000c73fd 100644
--- a/virttest/utils_net.py
+++ b/virttest/utils_net.py
@@ -1197,6 +1197,7 @@ def __init_openvswitch():
             if (not __ovs.check()):
                 raise Exception("Check of OpenVSwitch failed.")
         except Exception as e:
+            raise
             LOG.error("Host does not support OpenVSwitch: %s", e)
             __ovs = None
     return __ovs
```

**Reply by pevogam** [reply]
I see, thanks for the hints! So the stack trace is the following:
```
[stdlog] 2023-08-11 22:26:31,526 avocado.test ERROR|   File "/usr/lib/python3.10/site-packages/virttest/qemu_vm.py", line 3757, in destroy
[stdlog] 2023-08-11 22:26:31,526 avocado.test ERROR|     self._cleanup(free_mac_addresses)
[stdlog] 2023-08-11 22:26:31,526 avocado.test ERROR|   File "/usr/lib/python3.10/site-packages/virttest/qemu_vm.py", line 3688, in _cleanup
[stdlog] 2023-08-11 22:26:31,526 avocado.test ERROR|     self._del_port_from_bridge(nic)
[stdlog] 2023-08-11 22:26:31,526 avocado.test ERROR|   File "/usr/lib/python3.10/site-packages/virttest/qemu_vm.py", line 2840, in _del_port_from_bridge
[stdlog] 2023-08-11 22:26:31,526 avocado.test ERROR|     br_mgr, br_name = utils_net.find_current_bridge(nic.ifname)
[stdlog] 2023-08-11 22:26:31,526 avocado.test ERROR|   File "/usr/lib/python3.10/site-packages/virttest/utils_net.py", line 2005, in find_current_bridge
[stdlog] 2023-08-11 22:27:50,323 avocado.test ERROR|     ovs = __init_openvswitch()
[stdlog] 2023-08-11 22:27:50,323 avocado.test ERROR|   File "/usr/lib/python3.10/site-packages/virttest/utils_net.py", line 1195, in __init_openvswitch
[stdlog] 2023-08-11 22:27:50,323 avocado.test ERROR|     __ovs = factory(openvswitch.OpenVSwitchSystem)()
[stdlog] 2023-08-11 22:27:50,323 avocado.test ERROR|   File "/usr/lib/python3.10/site-packages/virttest/versionable_class.py", line 361, in factory
[stdlog] 2023-08-11 22:27:50,324 avocado.test ERROR|     return Manager(orig_cls.__module__).factory(orig_cls, *args, **kargs)[0]
[stdlog] 2023-08-11 22:27:50,324 avocado.test ERROR|   File "/usr/lib/python3.10/site-packages/virttest/versionable_class.py", line 308, in factory
[stdlog] 2023-08-11 22:27:50,324 avocado.test ERROR|     cls, cls_vn = self.factory(m_cls, *args, **kargs)
[stdlog] 2023-08-11 22:27:50,324 avocado.test ERROR|   File "/usr/lib/python3.10/site-packages/virttest/versionable_class.py", line 298, in factory
[stdlog] 2023-08-11 22:27:50,324 avocado.test ERROR|     if cl._is_right_ver(*args, **kargs):
[stdlog] 2023-08-11 22:27:50,324 avocado.test ERROR|   File "/usr/lib/python3.10/site-packages/virttest/openvswitch.py", line 255, in _is_right_ver
[stdlog] 2023-08-11 22:27:50,325 avocado.test ERROR|     version = cls.get_version()
[stdlog] 2023-08-11 22:27:50,467 avocado.test ERROR|   File "/usr/lib/python3.10/site-packages/virttest/openvswitch.py", line 173, in get_version
[stdlog] 2023-08-11 22:27:50,468 avocado.test ERROR|     path.find_command("ovs-vswitchd"))
[stdlog] 2023-08-11 22:27:50,468 avocado.test ERROR|   File "/usr/lib/python3.10/site-packages/avocado/utils/path.py", line 122, in find_command
[stdlog] 2023-08-11 22:27:50,468 avocado.test ERROR|     raise CmdNotFoundError(cmd, path_paths)
[stdlog] 2023-08-11 22:27:50,468 avocado.test ERROR| avocado.utils.path.CmdNotFoundError: Command 'ovs-vswitchd' could not be found in any of the PATH dirs: ['/bin', '/sbin', '/usr/bin', '/usr/lib64/ccache', '/usr/lib64/qt-3.3/bin', '/usr/libexec', '/usr/libexec/sdcc', '/usr/local/bin', '/usr/local/sbin', '/usr/sbin']
```
and is due to code in the env_process preprocessing hooks that destroy vms. Perhaps destroying a qemu vm is not supposed to lead to OpenVSwitch being checked for existence on each call?

**Reply by pevogam** [reply]
Hi @luckyh is this stracktrace still any useful to you?

**Reply by luckyh** [reply]
@pevogam thank you for providing this useful info! It's just hard for me to find some time working on this issue during the past busy months. Let's see if I can grab some time to dive deeper into it on Dec, thanks!

**Reply by pevogam** [reply]
Alright, let's see if we can make progress with this before the Avocado LTS release :+1:

**Reply by luckyh** [reply]
Hi @pevogam I'm not sure if the following would just be a partial fix to this issue but worth a try.
```patch
diff --git a/virttest/qemu_vm.py b/virttest/qemu_vm.py
index b58836be..fbc76b39 100644
--- a/virttest/qemu_vm.py
+++ b/virttest/qemu_vm.py
@@ -3700,7 +3700,7 @@ class VM(virt_vm.BaseVM):
             if nic.nettype == 'macvtap':
                 tap = utils_net.Macvtap(nic.ifname)
                 tap.delete()
-            elif nic.ifname:
+            elif nic.ifname and nic.nettype in ('bridge', 'network'):
                 port_mapping[nic.ifname] = nic
 
         if port_mapping:
```

And if we still can see the issue in the same context I guess something like this would be required.
```patch
diff --git a/virttest/utils_net.py b/virttest/utils_net.py
index e77b888f..c66760d0 100644
--- a/virttest/utils_net.py
+++ b/virttest/utils_net.py
@@ -1989,6 +1989,10 @@ def find_bridge_manager(br_name, ovs=None):
     return None
 
 
+def iface_has_master(ifname):
+    ...
+
+
 def find_current_bridge(iface_name, ovs=None):
     """
     Finds bridge which contains interface iface_name.
@@ -2000,6 +2004,8 @@ def find_current_bridge(iface_name, ovs=None):
     master = __bridge
     bridge = master.port_to_br(iface_name)
     if bridge is None:
+        if not iface_has_master(iface_name):
+            return(None, None)
         if ovs is None:
             ovs = __init_openvswitch()
         if ovs is not None:
```

**Comment by pevogam** [comment]
I tried both changes but still see the error, now from
```python
def find_bridge_manager(br_name, ovs=None):
    """
    Finds bridge which contain interface iface_name.

    :param br_name: Name of interface.
    :return: (br_manager) which contain bridge or None.
    """
    # find ifname in standard linux bridge.
    if br_name in __bridge.list_br():
        return __bridge
    if ovs is None:
        ovs = __init_openvswitch()
```
All of this makes me wonder if any of the above is even the right approach here since it seems most bridge management functions rely on optional `ovs` parameter that if not provided they struggle to initialize themselves and might still fail to do so with a side effect resulting in a boolean outcome. Perhaps we could choose one of these instead:
1) Make the `ovs` parameter optional in a simpler way - if provided it will be reused and if not provided no ovs will be initialized and assumed. OpenVSwitch can thus be initialized only once at the right user code location where it is actually meant to be used and/or required.
2) Reduce the logging verbosity or at least its level (current level is ERROR) when OpenVSwitch is not found since this is meant to be a simple check of availability and additional tools like it to help detect the bridge or bridge manager and not an actual error.
3) Add a switch like `REQUIRE_OVS` that can let us skip any ovs attempts:
```diff
+         if ovs is None and REQUIRE_OVS:
             ovs = __init_openvswitch()
```
What do you think?

**Reply by luckyh** [reply]
> All of this makes me wonder if any of the above is even the right approach here since it seems most bridge management functions rely on optional ovs parameter that if not provided they struggle to initialize themselves and might still fail to do so with a side effect resulting in a boolean outcome.

In fact, these issues were caused by a flaw of the model design in the vt qemu network control plane: instead of letting callers specify the driver/manager of the bridge devices it always executes an automatic detect, and as the result `ovs` will always be invoked if it assumes that `linux_bridge` was not the target for whatever reason.

We could fix the issues by changing the model design, however, the refactoring to be required will be complex and not go too fast I guess. On the other hand, IMO either a regular fix or a workaround against the current design would sound rational for us, that is,

- before invoking `ovs` at any place, do a check like the following:
  - if the given device is a bridge, test whether its driver is OpenVSwitch or not
    (e.g. by reading the NETLINK_INFO via `ip -json -details addr list $ifname`)
  - otherwise, test whether the given device associated with a bridge and the bridge's type is OpenVSwitch
    (similar as the above, as what I've suggested in the previous comment)

Or

- as you suggested we could workaround that by reducing the logging verbosity since those were just logs instead of exceptions.

I don't have a strong preference to the two options therefore I'd like to hear from you, thanks.

**Reply by pevogam** [reply]
Overall I think that adding additional checks for the OVS driver is only going to exacerbate the situation where each method is setting up OVS by itself even further by adding more repeated logic and reinforcing the design so I would likely rather stick with de-escalating the logging level to DEBUG or INFO level and suggest the design gets improved from its current state to not have to log such lines for users not using these drivers in the future. Does that sound good?

**Reply by luckyh** [reply]
Ok, I just issued https://github.com/avocado-framework/avocado-vt/pull/3821 per our discussion, thanks!

**Reply by pevogam** [reply, marked as answer]
Alright, closing this as the suggested pull request was merged, I do hope we can improve on the overall design in the future though.
